x/mm2/keeper/keeper.go

```diff
@@ -0,0 +1,203 @@
+	)
+
+	return &Keeper{
+		cdc:         cdc,
+		authority:   authority,
+		markets:     collections.NewMap(sb, types.MarketsPrefix, "markets", types.TickersCodec, codec.CollValue[types.Market](cdc)),
+		lastUpdated: collections.NewItem[uint64](sb, types.LastUpdatedPrefix, "last_updated", types.LastUpdatedCodec),
+		params:      params,
+	}
+}
+
+// SetLastUpdated sets the lastUpdated field to the current block height.
+func (k *Keeper) SetLastUpdated(ctx sdk.Context, height uint64) error {
+	return k.lastUpdated.Set(ctx, height)
+}
+
+// GetLastUpdated gets the last block-height the market map was updated.
+func (k *Keeper) GetLastUpdated(ctx sdk.Context) (uint64, error) {
+	return k.lastUpdated.Get(ctx)
+}
+
+// GetMarket returns a market from the store by its currency pair string ID.
+func (k *Keeper) GetMarket(ctx sdk.Context, tickerStr string) (types.Market, error) {
+	return k.markets.Get(ctx, types.TickerString(tickerStr))
+}
+
+// GetAllMarkets returns the set of Market objects currently stored in state.
+func (k *Keeper) GetAllMarkets(ctx sdk.Context) ([]types.Market, error) {
+	iter, err := k.markets.Iterate(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	markets, err := iter.Values()
+	if err != nil {
+		return nil, err
+	}
+	return markets, err
+}
+
+// GetAllMarketsMap returns the set of Market objects currently stored in state
+// as a map[TickerString] -> Markets.
+func (k *Keeper) GetAllMarketsMap(ctx sdk.Context) (map[string]types.Market, error) {
+	iter, err := k.markets.Iterate(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	keyValues, err := iter.KeyValues()
+	if err != nil {
+		return nil, err
+	}
+
+	m := make(map[string]types.Market, len(keyValues))
+	for _, keyValue := range keyValues {
+		m[string(keyValue.Key)] = keyValue.Value
+	}
+
+	return m, nil
+}
+
+// createMarket initializes a new Market.
+// The Ticker.String corresponds to a market, and must be unique.
+func (k *Keeper) createMarket(ctx sdk.Context, market types.Market) error {
+	// Check if Ticker already exists for the provider
+	alreadyExists, err := k.markets.Has(ctx, types.TickerString(market.Ticker.String()))
+	if err != nil {
+		return err
+	}
+	if alreadyExists {
+		return types.NewMarketAlreadyExistsError(types.TickerString(market.Ticker.String()))
+	}
+	// Create the config
+	return k.markets.Set(ctx, types.TickerString(market.Ticker.String()), market)
+}
+
+// updateMarket updates a Market.
+// The Ticker.String corresponds to a market, and exist unique.
+func (k *Keeper) updateMarket(ctx sdk.Context, market types.Market) error {
+	// Check if Ticker already exists for the provider
+	alreadyExists, err := k.markets.Has(ctx, types.TickerString(market.Ticker.String()))
+	if err != nil {
+		return err
+	}
+	if !alreadyExists {
+		return types.NewMarketDoesNotExistsError(types.TickerString(market.Ticker.String()))
+	}
+	// Create the config
+	return k.markets.Set(ctx, types.TickerString(market.Ticker.String()), market)
```

Comment:
No validation needs to happen on which fields are being updated?

Reply:
What kind of validation do you mean?

Reply:
I guess we don't need any? My initial thought was something along the lines of "which fields don't we want to make mutable". But I guess just the ticker's currency pair.